rfcs/image-scanning.md

```diff
@@ -0,0 +1,584 @@
+ORAS client, in fact, supports extending the mime types via [config](https://github.com/deislabs/oras/pull/56)  
+Open Policy Agent also [opted](https://github.com/open-policy-agent/opa/issues/1413) for the definition of well-known mime types:
+
+Given the above assumptions regarding mime types, we can proceed implementing the storage of the vulnerabilities report as follow:
+
+* Step 1: Store the security scanning report JSON response as json/yaml `application/vnd.oci.vuln.report.v1+[json|yaml]` file and calculate its sha256 hash
+* Step 2: Write the following `manifest.json`:  
+```json
+
+{
+  "schemaVersion": 2,
+  "config": {
+    "mediaType": "vnd.unknown.config.v1+json"
+  },
+  "layers": [
+    {
+      "mediaType": "application/vnd.oci.vuln.report.v1+json",
+      "size": 32654,
+      "digest": "sha256:9834876dcfb05cb167a5c24953eba58c4ac89b1adf57f28f2f9d09af107ee8f0"
+    }    
+  ]
+}
+```
+We do not need a config manifest for the vulnerability scanning report, so the mediaType is set to: `vnd.unknown.config.v1+json`  
+See Azure container [documentation](https://docs.microsoft.com/en-us/azure/container-registry/container-registry-oci-artifacts)  
+
+* Step 3: Add an entry to the `index` file of the image which was scanned.  
+This will also serve as a quick way to check whether an image/artifact was already scanned 
+(there is no info about the date and time of the scanning though, so we will need to lookup the artifact as well in case it's needed).    
+Example:
+```json
+{
+  "schemaVersion": 2,
+  "manifests": [
+    {
+      "mediaType": "application/vnd.oci.image.manifest.v1+json",
+      "size": 7143,
+      "digest": "sha256:e692418e4cbaf90ca69d05a66403747baa33ee08806650b51fab815ad7fc331f",
+      "platform": {
+        "architecture": "ppc64le",
+        "os": "linux"
+      }
+    },
+    {
+      "mediaType": "application/vnd.oci.vuln.meta.v1+json",
+      "size": 7143,
+      "digest": "sha256:e692418e4cbaf90ca69d05a66403747baa33ee08806650b51fab815ad7fc331f",
+      "platform": {
+        "architecture": "ppc64le",
+        "os": "linux"
+      }
+    },
+    {
+      "mediaType": "application/vnd.oci.image.manifest.v1+json",
+      "size": 7682,
+      "digest": "sha256:5b0bcabd1ed22e9fb1310cf6c2dec7cdef19f0ad69efa1f392e94a4333501270",
+      "platform": {
+        "architecture": "amd64",
+        "os": "linux"
+      }
+    },
+    {
+      "mediaType": "application/vnd.oci.vuln.meta.v1+json",
+      "size": 7682,
+      "digest": "sha256:5b0bcabd1ed22e9fb1310cf6c2dec7cdef19f0ad69efa1f392e94a4333501270",
+      "platform": {
+        "architecture": "amd64",
+        "os": "linux"
+      }
+    }    
+  ],
+  "annotations": {
+    "com.example.key1": "value1",
+    "com.example.key2": "value2"
+  }
+}
+``` 
+
+#### Trow WebHook
+Once an image is found to have vulnerabilities we can allow pushing this information out to third party HTTP/s end points.  
```

Comment:
I like this. Does Harbor do the same thing? What are the use cases?

I wonder if rather than send out the report it would be better just to say "this image has an updated vulnerability report" and leave it to them to go and pull the report if interested.

Reply:
I am not aware of Harbor doing this, but I find it useful for simplifying third party integrations and in general to build SIEM systems. 
We can either send the full report (which as long as the client implements the signature verification it's secure) or as you proposed just send a simpler JSON (this image has an updated vulnerability report).  
The reason why I was suggesting to send the full report is that, as a developer, I prefer to keep the integration steps to a minimum. So basically if we send only the information regarding `this image has an updated vulnerability report` then as a developer I will need to:
- verify the signature of the webhook (which requires a shared secret key between both the client and the server)
- add API key to my http client to authenticate the next request I need to make to Trow API to pull the report
So I end up with having to maintain 2 secrets.

Reply:
Yeah, it might be easiest to send the whole report. I'm just thinking they might be quite large in some cases, but it's text that compresses well.

Regarding security, is it enough to say that if you can pull the image you can get the security report? After all they could run the security scan themselves.